src/coreclr/src/pal/src/thread/thread.cpp

```diff
@@ -1763,6 +1732,42 @@ CPalThread::ThreadEntry(
         goto fail;
     }
 
+#if HAVE_SCHED_GETAFFINITY && HAVE_SCHED_SETAFFINITY
+    // Threads inherit their parent's affinity mask on Linux. This is not desired, so we reset
+    // the current thread's affinity mask to the mask of the current process.
+    //
+    // Typically, we would use pthread_attr_setaffinity_np() and have pthread_create() create the thread with the specified
+    // affinity. At least one implementation of pthread_create() following a pthread_attr_setaffinity_np() calls
+    // sched_setaffinity(<newThreadPid>, ...), which is not allowed under Snap's default strict confinement without manually
+    // connecting the process-control plug. To work around that, have the thread set the affinity after it starts.
+    // sched_setaffinity(<currentThreadPid>, ...) is also currently not allowed, only sched_setaffinity(0, ...).
+    // pthread_setaffinity_np(pthread_self(), ...) seems to call sched_setaffinity(<currentThreadPid>, ...) in at least one
+    // implementation, and does not work. Use sched_setaffinity(0, ...) instead. See the following for more information:
+    // - https://github.com/dotnet/runtime/pull/38795
+    // - https://github.com/dotnet/runtime/issues/1634
+    // - https://forum.snapcraft.io/t/requesting-autoconnect-for-interfaces-in-pigmeat-process-control-home/17987/13
+
+    CPU_ZERO(&cpuSet);
+
+    st = sched_getaffinity(gPID, sizeof(cpu_set_t), &cpuSet);
```

Comment:
A nit - zero instead of gPID should work here too.

Reply:
Based on the docs I'm not completely sure that they're the same thing. The previous intention was to get the affinity of the process (or perhaps the thread corresponding with `gPID`) and apply the same affinity to the new thread. The docs for `sched_getaffinity` only talks about "process" and not "thread", but I wonder if there is ambiguity there since threads have their own pid the the same way as processes.

Reply:
You are actually right, my mistake. The goal of this code is to set the affinity to be the same as the affinity of the main thread (which is identified by the same PID as the process). If we used zero here, the operation would be effectively noop, as it would just get the affinity of the current thread and then set it to the same value.

Reply:
I was looking at https://linux.die.net/man/2/sched_getaffinity, it does mention a bit about threads:

`The affinity mask is actually a per-thread attribute that can be adjusted independently for each of the threads in a thread group. The value returned from a call to gettid(2) can be passed in the argument pid. Specifying pid as 0 will set the attribute for the calling thread, and passing the value returned from a call to getpid(2) will set the attribute for the main thread of the thread group. (If you are using the POSIX threads API, then use pthread_setaffinity_np(3) instead of sched_setaffinity().)`

According to https://man7.org/linux/man-pages/man2/sched_getaffinity.2.html, which talks about threads, it's pretty clear from:

`If pid is zero, then the mask of the calling thread is returned.`

It doesn't sound like it would mean the same thing.

Reply:
@jdstrand, is `sched_getaffinity` restricted in process-id usage, too? i.e. would the usage above fall foul of the same situation that we have with `sched_setaffinity` that this PR aims to fix?

Reply:
It does not seem to be restricted, only `sched_setaffinity` is failing

Reply:
> @jdstrand, is `sched_getaffinity` restricted in process-id usage, too? i.e. would the usage above fall foul of the same situation that we have with `sched_setaffinity` that this PR aims to fix?

sched_getaffinity is part of the default template (since it doesn't allow interfering with other snaps and the small info leak isn't worth the developer friction of putting it in another interface).

Reply:
Awesome :-) thanks for this!